File: Mecanismos_de_Sync/Barreira.java
package Mecanismos_de_Sync;
import java.util.concurrent.Semaphore;

public class Barreira {

    private int count;
    private Semaphore semaphore;
    private Semaphore mutex;
    private int bar;

    public Barreira(int bar) {
        this.count = 0;
        this.semaphore = new Semaphore(0);
        this.mutex = new Semaphore(1);
        this.bar = bar;
    }

    public void waitt(){
        try {
            mutex.acquire();
            count++;
            int c = count;            
            mutex.release();

            if(c == bar){
                semaphore.release();
            }

            semaphore.acquire();   
            semaphore.release();

        } catch (Exception e) {
        }
   
    }

}

File: Mecanismos_de_Sync/Barreira.py
from threading import Semaphore

n = int(input())

class Barrier: 
    def __init__(self, n) -> None:
        self.n = n
        self.count = 0
        self.s1 = Semaphore(0)
        self.mutex = Semaphore(1)

    def wait (self):
        self.mutex.acquire()
        self.count += 1
        count = self.count
        print(f'c: {self.count}')
        print(f'co: {count}')
        self.mutex.release()    

        if (count ==  self.n):
            print("Estou livre")
            self.s1.release()

        self.s1.acquire()
        self.s1.release()
    